File: python/322_Coin Change.py
#!/usr/bin/env python2
#coding=utf8

class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        memory = [float('inf')] * (amount + 1)
        memory[0] = 0
        for i in range(1, amount+1):
            for c in coins:
                if i >= c:
                    memory[i] = min(memory[i], 1 + min([memory[i-c] for c in coins if i - c >= 0] + [float('inf')]))
        if memory[-1] == float('inf'):
            return -1
        else:
            return memory[-1]

s = Solution()
print s.coinChange([70,177,394,428,427,437,176,145,83,370],7582)

# https://discuss.leetcode.com/topic/32589/fast-python-bfs-solution
# https://discuss.leetcode.com/topic/35720/easy-to-understand-recursive-dp-solution-using-java-with-explanations

File: java/src/brand_new/_322_Coin_Change.java
package brand_new;

import java.util.Arrays;

public class _322_Coin_Change {
	public int coinChange(int[] coins, int amount) {
		if (amount == 0)
			return 0;
		int[] M = new int[amount + 1];
		for (int i = 1; i < amount + 1; i++)
			M[i] = amount + 1;

		Arrays.sort(coins);

		for (int i = 0; i < coins.length; i++) {
			for (int j = coins[i]; j < amount + 1; j++) {
				M[j] = Math.min(M[j - coins[i]] + 1, M[j]);
			}
		}
		
		
		return M[amount] > amount ? -1 : M[amount];
	}

	public static void main(String[] args) {
		System.out.println(new _322_Coin_Change().coinChange(new int[] { 1, 2,
				5 }, 11));
	}

}
